datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -557,570 +971,510 @@ fn make_renamed_schema(
+    };
+
+    // ref: https://substrait.io/relations/physical_relations/#exchange-types
+    let partitioning_scheme = match exchange_kind {
+        ExchangeKind::ScatterByFields(scatter_fields) => {
+            let mut partition_columns = vec![];
+            let input_schema = input.schema();
+            for field_ref in &scatter_fields.fields {
+                let column = from_substrait_field_reference(field_ref, input_schema)?;
+                partition_columns.push(column);
             }
-            Err(e) => not_impl_err!("Invalid set operation type {}: {e}", set.op),
-        },
-        Some(RelType::ExtensionLeaf(extension)) => {
-            let Some(ext_detail) = &extension.detail else {
-                return substrait_err!("Unexpected empty detail in ExtensionLeafRel");
-            };
-            let plan = state
-                .serializer_registry()
-                .deserialize_logical_plan(&ext_detail.type_url, &ext_detail.value)?;
-            Ok(LogicalPlan::Extension(Extension { node: plan }))
-        }
-        Some(RelType::ExtensionSingle(extension)) => {
-            let Some(ext_detail) = &extension.detail else {
-                return substrait_err!("Unexpected empty detail in ExtensionSingleRel");
-            };
-            let plan = state
-                .serializer_registry()
-                .deserialize_logical_plan(&ext_detail.type_url, &ext_detail.value)?;
-            let Some(input_rel) = &extension.input else {
-                return substrait_err!(
-                    "ExtensionSingleRel doesn't contains input rel. Try use ExtensionLeafRel instead"
-                );
-            };
-            let input_plan = from_substrait_rel(state, input_rel, extensions).await?;
-            let plan =
-                plan.with_exprs_and_inputs(plan.expressions(), vec![input_plan])?;
-            Ok(LogicalPlan::Extension(Extension { node: plan }))
+            Partitioning::Hash(partition_columns, exchange.partition_count as usize)
         }
-        Some(RelType::ExtensionMulti(extension)) => {
-            let Some(ext_detail) = &extension.detail else {
-                return substrait_err!("Unexpected empty detail in ExtensionSingleRel");
-            };
-            let plan = state
-                .serializer_registry()
-                .deserialize_logical_plan(&ext_detail.type_url, &ext_detail.value)?;
-            let mut inputs = Vec::with_capacity(extension.inputs.len());
-            for input in &extension.inputs {
-                let input_plan = from_substrait_rel(state, input, extensions).await?;
-                inputs.push(input_plan);
-            }
-            let plan = plan.with_exprs_and_inputs(plan.expressions(), inputs)?;
-            Ok(LogicalPlan::Extension(Extension { node: plan }))
+        ExchangeKind::RoundRobin(_) => {
+            Partitioning::RoundRobinBatch(exchange.partition_count as usize)
         }
```

Comment:
These Extension relation handling code has been fully removed as that must be provided by users.

Reply:
Would it make sense to keep these in the default implementation though? Just as in the producer we do produce these types (https://github.com/apache/datafusion/blob/b8e435f3e38f1f8d2d8f24c3da3c2b21e76b700b/datafusion/substrait/src/logical_plan/producer.rs#L596), and I guess it'd be nice if the roundtrip tests that don't require special setup on producer wouldn't require special setup for consumer either.

That's not a strictly held opinion.

Reply:
> Would it make sense to keep these in the default implementation though?

I think it would be reasonable to include these in the `DefaultSubstraitConsumer`. That would play nice with roundtrip tests, and it might be a reasonable default to assume that extension relations are `UserDefinedLogicalNode`s.

I'll add these back in.